ddtrace/tracer/sampler.go

```diff
@@ -185,42 +186,51 @@ func newRulesSampler(rules []SamplingRule) *rulesSampler {
 	}
 }
 
-// appliedSamplingRules validates the user-provided rules and returns an internal representation.
-// If the DD_TRACE_SAMPLING_RULES environment variable is set, it will replace the given rules.
-func appliedSamplingRules(rules []SamplingRule) []SamplingRule {
+// samplingRulesFromEnv parses sampling rules from the DD_TRACE_SAMPLING_RULES
+// environment variable.
+func samplingRulesFromEnv() ([]SamplingRule, error) {
 	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
-	if rulesFromEnv != "" {
-		rules = rules[:0]
-		jsonRules := []struct {
-			Service string      `json:"service"`
-			Name    string      `json:"name"`
-			Rate    json.Number `json:"sample_rate"`
-		}{}
-		err := json.Unmarshal([]byte(rulesFromEnv), &jsonRules)
+	if rulesFromEnv == "" {
+		return nil, nil
+	}
+	jsonRules := []struct {
+		Service string      `json:"service"`
+		Name    string      `json:"name"`
+		Rate    json.Number `json:"sample_rate"`
+	}{}
+	err := json.Unmarshal([]byte(rulesFromEnv), &jsonRules)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing DD_TRACE_SAMPLING_RULES: %v", err)
+	}
+	rules := make([]SamplingRule, 0, len(jsonRules))
+	var errStr string
+	for _, v := range jsonRules {
+		if v.Rate == "" {
+			errStr += "rate not provided "
+			continue
+		}
+		rate, err := v.Rate.Float64()
 		if err != nil {
-			log.Warn("error parsing DD_TRACE_SAMPLING_RULES: %v", err)
-			return nil
+			errStr += fmt.Sprintf("invalid rate: %v ", err)
+			continue
 		}
-		for _, v := range jsonRules {
-			if v.Rate == "" {
-				log.Warn("error parsing rule: rate not provided")
-				continue
-			}
-			rate, err := v.Rate.Float64()
-			if err != nil {
-				log.Warn("error parsing rule: invalid rate: %v", err)
-				continue
-			}
-			switch {
-			case v.Service != "" && v.Name != "":
-				rules = append(rules, NameServiceRule(v.Name, v.Service, rate))
-			case v.Service != "":
-				rules = append(rules, ServiceRule(v.Service, rate))
-			case v.Name != "":
-				rules = append(rules, NameRule(v.Name, rate))
-			}
+		switch {
+		case v.Service != "" && v.Name != "":
+			rules = append(rules, NameServiceRule(v.Name, v.Service, rate))
+		case v.Service != "":
+			rules = append(rules, ServiceRule(v.Service, rate))
+		case v.Name != "":
+			rules = append(rules, NameRule(v.Name, rate))
 		}
 	}
+	if errStr != "" {
+		return rules, fmt.Errorf("error(s) parsing DD_TRACE_SAMPLING_RULES: %s", errStr)
+	}
```

Comment:
Can we drop this and instead just return on the first error? Most parsers work like that. No need to continue collecting and merging error strings.

Reply:
We can't return on the first error, since that would break existing expectations that the reader continue to try the other rules.

Reply:
Ok I see. Thanks for clarifying. In that case, it might be easier to have just one error. Or, alternatively, you could make them nice and tidy. But please add tests too, so we see what they look like.

Reply:
We'll make them look pretty. 😄 

I would prefer to include as many error messages as possible. If we know about errors, we should report them to the user so they can address them.